### PR TITLE
[JENKINS-57519] JavaScript: switch to let/const

### PR DESCRIPTION
**Pull request by jsoref**
This is for discussion.

Notes:
* All items were identified by IntelliJ (i.e. this is not a global replace)
* Individual commits are mostly self contained
* In general, most commits should not change program behavior at all

### Proposed changelog entries

* `Internal:` JS: change from `var` to `let`

### Submitter checklist

- [ ] JIRA issue is well described
- [ ] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### CODE REVIEW

**Review by oleg-nenashev — DISMISSED** [review]
okay? CC @halkeye

**Comment by halkeye** [comment]
None of these should change behaviour, assuming they are already defined like the description says. so no objection from me.

If the code gets transpiled, then changing them to const/let would be better as then each scope would be accurately defined.

So yea, no objection

**Comment by jsoref** [comment]
It doesn't look like `let` is being used in `core/src/main/resources/hudson`, so, while that's something that might be interesting, I'd rather someone who has a better grasp of what Jenkins's client matrix looks like and can apply it more uniformly across the project... (Possibly even w/ a JEP?)

**Comment by Wadeck** [comment]
TL;DR: I am against this change as it reduces the readability, but not blocking if other finds an interest.

Personally I am opposed to this change as it reduce the readability of the code. Any good JavaScript developer knows the var scoping troubles, but having no declaration of the variable means that the variable was either declared before, or is a global variable. As the second option is very bad, the people that reads that code, should consider validating the variable is declared before. That leads to lost of time, due to this simplification.

@jsoref if you wanted to have info on the browser matrix compatibility it's there: https://wiki.jenkins.io/display/JENKINS/Browser+Compatibility+Matrix (quite old but only doc I think).

In addition, when (if?) we will change the `var` into `let` / `const`, the refactoring will be harder after this PR.

**Comment by daniel-beck** [comment]
I think @rtyler had some trouble with our UI testing frameworks when modernizing some JS a while back, so I wouldn't be surprised if our tests would start failing with this.

**Comment by jsoref** [comment]
@daniel-beck I wouldn't be surprised.

@rtyler: this looks like infra failing, can someone please kick it?
 https://ci.jenkins.io/blue/organizations/jenkins/Core%2Fjenkins/detail/PR-4025/4/pipeline

Would I need to run ath manually, or does the automatic system give us sufficient coverage?

**Comment by daniel-beck** [comment]
I think the build result is reasonable and we don't need to wait for the Windows build to pass.

**Comment by Wadeck** [comment]
For your information, `let` vs `const`, `let` is less supported (esp. by IE). [let support](https://caniuse.com/#search=let) vs [const support](https://caniuse.com/#search=const).

⚠️ Some of the `var` => `let` in this PR are misleading for the next readers. Some variables should just using `const` instead of `let`. I am fine with the ones in the `for` loops, but for regular variables, I would just say "no". 90% of the time a `var` is actually a constant (except for the loops).

No blocker but that will add more difficulties for the next ones, for no real value from my PoV.

**Comment by jsoref** [comment]
@Wadeck: I'm ok w/ switching... -- switched...

**Comment by varyvol** [comment]
@jenkinsci/core-pr-reviewers some more reviews could be helpful here.

**Comment by rsandell** [comment]
Note on browser compatibility https://wiki.jenkins.io/display/JENKINS/Browser+Compatibility+Matrix seems to be ok?

**Review by halkeye — APPROVED** [review]
Since we are already using const/let Its probably fine. 

But as I said earlier using them does have the risk of breaking older browsers.

I know at least ie 10, and possibly 11, and safari from that era.

So there's risk and I don't really understand the reward of doing this

**Comment by Wadeck** [comment]
> Since we are already using const/let Its probably fine.

For the precision, there are no such usage in "core" ;)

**Review by halkeye — CHANGES_REQUESTED** [review]
> For the precision, there are no such usage in "core" ;)

oh then yea, i change my mind, i don't think there's any benefit to this change, and we reduce who will be able to access jenkins, so I don't think its a good plan

**Comment by fcojfernandez** [comment]
After reviewing this PR, it seems there is no approvals and the community is against this change, so we've just added the `proposed-for-close` label.

**Comment by fcojfernandez** [comment]
It's been a long time since this PR has been proposed for close

**Comment by fqueiruga** [comment]
We can reconsider these changes in the future if/when we add a transpiler. We would be able to use ES6 features then.
